### ISSUE TITLE
Anchor names for non-ASCII strings

### ISSUE THREAD

**Issue by Yenya** [briandfoy/text-multimarkdown#6]
Hello!

I would like to have more readable labels (as produced by `_Header2Label()`). There are two ways of improving the generated labels:

1. Both pandoc(1) and Github Markdown implementations support the following syntax to set the anchor name (id/label) explicitly:

```
## Some title {#completely-different-label}
```

2. I am from non-English language environment, and we use Latin with lots of diacritics. It would be nice if labels can be generated by transliterating Latin characters with diacritics to ASCII instead of stripping them altogether. For example, this markdown source:
```
## Naîve Citroën driver from Føroyar
```
would lead to this HTML code:
```
<h2 id="naive-citroen-driver-from-foroyar">Naîve Citroën driver from Føroyar</h2>
```

It can be done using `Text::Iconv`:
```
use Text::Iconv;
use feature 'state';

sub _Header2Label {
     ...
     state $conv //= Text::Iconv->new('utf-8', 'us-ascii//TRANSLIT');
     $label = $conv->convert($label);    # transliterate to ASCII
     $label =~ s/[^A-Za-z0-9:_.-]//g;        # Strip illegal characters
}
```

What do you think about this? Should I create a pull request at least for the second part? If so - is it OK to use a `state` variable in this module? Should the labels remain backward-compatible and use transliteration as a new configuration option?

Thanks!

**Comment by Yenya** [comment]
OK, I'll use `our` instead of `state` then.

ASCII-only labels will be compatible, of course. OK, I'll try to make a pull req.

**Comment by Yenya** [comment]
Brian,

I have been looking at the specifications about what characters are permitted in the ID of a HTML tag. It seems that HTML5 allows any characters, HTML4 allows only ASCII (with further restrictions, such as that ID cannot start with a digit or two hyphens), and CSS2 is somewhere in between - it explicitly allows any character greater than U+00A0 to be part of an ID, though.

Pandoc keeps Latin letters with diacritics in the `id=` labels when converting Markdown to HTML. It just transforms them to lower case. So I wonder wheter a simple `s/\W+/-/g` would do the job as well, instead of depending on `Text::Iconv`.

What do you think? How conservative you want `Text::MultiMarkdown` to be when generating labels?

Thanks!

-Yenya

**Comment by briandfoy** [comment]
I started looking into this, then just made the change. Take a look at the code in the repo and the `transliterate_ids` and `unicode_ids` settings.

There are is also a new `to_html` method which is the same thing with a better name, and `multimarkdown_to_html` function which is also the same thing with a better name.

**Comment by Yenya** [comment]
Interesting approach. Looks ok, thanks!

As for `## Some title {#completely-different-label}` in my post above - it might make sense if the `*_id_handler` functions can return _both_ the label and the modified text of the label. That way it could be possible to return `completely-different-label` as id, and `Some title` as the text.

**Comment by briandfoy** [comment]
That will have to be another feature for another day.

**Comment by briandfoy** [comment]
I've released 1.002_01 and I'll wait a couple weeks for CPAN Testers to do their work.

There will be a change in the next version, where `transliterate_ids` will be `transliterated_ids` (adjective instead of verb). But, I think all the stuff is in place and I was able to get it to work with v5.8 in GitHub Actions. This required quite a bit of fiddling to figure out and once I got it running had to take out some v5.10 features I had snuck in.

Play around with it a bit and see if you can tolerate the changes.

**Comment by briandfoy** [comment]
@Yenya I've moved the `## Some title {#completely-different-label}` to #7. If you want to look at how that might work, comment there.

**Comment by briandfoy** [comment]
This is only held up by me looking at the CPAN Testers reports where the tests fail if HTML::Tidy is installed. However, the tests that failed say that they are going to fail. I think most of the problem is how tidy has changed some things and how other decisions have evolved in the past 10 years. But, also, some of the things are just broken I think.

**Comment by briandfoy** [comment]
I've released this as 1.003.
